### PR TITLE
Create QA plan 

### PR DESCRIPTION
**Pull request by mpoke**

### CODE REVIEW

File: docs/quality_assurance.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+- Correct software engineering practices (e.g., error handling, serialization, deserialization).
+- The correct integration with IBC (i.e., [ibc-go](https://github.com/cosmos/ibc-go/tree/v3.0.0)).
+- The correct integration with Cosmos SDK (i.e., [cosmos-sdk](https://github.com/cosmos/cosmos-sdk/tree/v0.45.6)).
+- The correctness of the provider chain, i.e., the provider CCV module does not break the liveness or the safety of the provider chain. 
+- The correctness of the Interchain Security protocol, i.e., the protocol follows the [specification](https://github.com/cosmos/ibc/blob/master/spec/appics-028-cross-chain-validation/README.md).
+- The correctness of the consumer chain, i.e., both liveness or safety hold.
+
+> TODO decide what should be covered by the audit and the priority
+
+## Software Engineering
```
Comment (mpoke) [inline docs/quality_assurance.md:23]: @danwt @sainoe Is there something else that we should check for in this section?

File: docs/quality_assurance.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+## Integration with Cosmos SDK
+
+A prerequisite of the code review is to open a PR with all the [SDK changes](https://github.com/cosmos/cosmos-sdk/tree/interchain-security-rebase) needed by Interchain Security.
+
+| Concern | Code Review | Unit Testing | Diff. testing | Testnet | Protocol audit |
+| ------- | ----------- | ------------ | ------------- | ------- | ----- |
+| Changes to staking module | `TODO` (sdk team) | `Low coverage` | `??` | `TODO` | `NA` |
+| Changes to slashing module | `TODO` (sdk team) | `??` | `??` | `TODO` | `NA` |
+| Changes to evidence module | `TODO` (sdk team) | `??` | `??` | `TODO` | `NA` |
```
Comment (danwt) [inline docs/quality_assurance.md:61]: I didn't even know we had changes to slashing or evidence! Did this happen yet?

**Review by danwt — COMMENTED** [review]
I wonder if we can be more precise about the statuses. `TODO` can mean anything from 'It's scheduled for this week' to 'maybe one day in the future, probably...' . Can we instead use terms like `Scheduled` for something which is concretely planned and `Left for future` for things which we would like but that aren't actually scheduled?

Feel free to suggest different terms, I just care about having precision.

EDIT: 

Maybe we can have `Done`, `Scheduled`, `In future`, `Not planned`.

File: docs/quality_assurance.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+| ------- | ----------- | ------------ | ------------- | ------- | ----- |
+| Every validator set on any consumer chain MUST either be or have been a validator set on the provider chain. | `TODO` | `NA` | `??` | `TODO` | `TODO` |
+| Every consumer chain receives the same sequence of `ValidatorSetChangePacket`s in the same order. | `TODO` | `NA` | `??` | `TODO` | `TODO` |
+
+---
+
+| Concern re. *Bond-Based Consumer Voting Power* | Code Review | Unit Testing | Diff. testing | Testnet | Protocol audit |
+| ------- | ----------- | ------------ | ------------- | ------- | ----- |
+| A power increase of a validator `val` on a consumer chain can be only due to <br /> - a `Delegate()` / `Redelegate()` to `val` on provider <br /> - `val` joining the provider validator set (another validator leaving the set) | `TODO` | `TODO` | `??` | `TODO` | `TODO` |
+| A power decrease of a validator `val` on a consumer chain can be only due to  <br /> - an `Undelegate()` / `Redelegate()` from `val` on provider <br /> - `val` gets slash on the provider chain <br /> - another validator joining the provider validator set (`val` leaving the set) | `TODO` | `TODO` | `??` | `TODO` | `TODO` |
```
Comment (danwt) [inline docs/quality_assurance.md:117]: I think we should reword these two concerns. It's not clear exactly what the issue is and they aren't really testable.

Maybe something like

- 'A positive power for validator `val` on a consumer chain at height `h1` is always corresponds to an equal positive power on the provider at height `h0`, `h0 < h1` under the `<` imposed by packet delivery.

ect

But even this is getting into the weeds. Should just state the need in plain English here and link to the fully fleshed property the spec?

File: docs/quality_assurance.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+| ICS-20 channel creation | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+| ICS-20 transfer | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+
+> @AdityaSripal Is there something else that we should check for? 
+
+## Integration with Cosmos SDK
+
+A prerequisite of the code review is to open a PR with all the [SDK changes](https://github.com/cosmos/cosmos-sdk/tree/interchain-security-rebase) needed by Interchain Security.
+
+| Concern | Code Review | Unit Testing | Diff. testing | Testnet | Audit |
```
[the remark this answers is not in the document]
Comment (sainoe) [reply docs/quality_assurance.md]: Everything is here for me.

File: docs/quality_assurance.md (hunk 2 above)
Comment (sainoe) [reply docs/quality_assurance.md:61]: Yes it did. The changes are mostly located in `x/slashing/keeper/infractions.go` and `x/evidence/keeper/infraction.go` and allow the CCV module to not necessarily have (for consumer chains) validator public keys and operator addresses stored in states.

File: docs/quality_assurance.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+| Create IBC clients | `TODO` (ibc-go team) | `Done` | `??` | `TODO` | `NA` |
+| Getting consumer `UnbondingPeriod` from IBC client | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+| Create CCV channel (handshake) | `TODO` (ibc-go team) | `Done` | `??` | `TODO` | `TODO` |
+| Sending IBC packets <br /> - see `x/ccv/utils/utils.go:SendIBCPacket()` | `TODO` (ibc-go team) | `Done` | `??` | `TODO` | `NA` |
+| Handling acknowledgments | `TODO` (ibc-go team) | `Low coverage` | `??` | `TODO` | `NA` |
+| Handling timeouts | `TODO` (ibc-go team) | `Low coverage` | `??` | `TODO` | `NA` |
+| Handling IBC client expiration | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+| ICS-20 channel creation | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+| ICS-20 transfer | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
+
```
Comment (sainoe) [inline docs/quality_assurance.md:52]: | Changes in IBC-GO testing suite | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |

Comment (sainoe) [inline docs/quality_assurance.md:52]:
```suggestion
| Changes in IBC-GO testing suite | `TODO` (ibc-go team) | `??` | `??` | `TODO` | `NA` |
```

File: docs/quality_assurance.md (hunk 3 above)
Comment (mpoke) [reply docs/quality_assurance.md]: What about https://github.com/cosmos/interchain-security/pull/215/commits/528262b34fcff141cdc4e2483b289901e871f9a3?

Comment (danwt) [reply docs/quality_assurance.md]: Yes, it's much clearer, nice 👍

File: docs/quality_assurance.md (hunk 2 above)
Comment (danwt) [reply docs/quality_assurance.md:61]: Great, thanks.

File: docs/quality_assurance.md [hunk 6]
```diff
@@ -41,26 +41,26 @@ IBC packets:
...
 ## Integration with Cosmos SDK
 
 A prerequisite of the code review is to open a PR with all the [SDK changes](https://github.com/cosmos/cosmos-sdk/tree/interchain-security-rebase) needed by Interchain Security.
 
 | Concern | Code Review | Unit Testing | Diff. testing | Testnet | Protocol audit |
 | ------- | ----------- | ------------ | ------------- | ------- | ----- |
-| Changes to staking module | `TODO` (sdk team) | `Low coverage` | `Partially done` | `TODO` | `NA` |
-| Changes to slashing module | `TODO` (sdk team) | `??` | `Not planned` | `TODO` | `NA` |
-| Changes to evidence module | `TODO` (sdk team) | `??` | `Not planned` | `TODO` | `NA` |
+| Changes to staking module | `Scheduled` (sdk team) | `Low coverage` | `WIP` | `Scheduled` | `NA` |
```
Comment (danwt) [inline docs/quality_assurance.md:61]: Hey, so in this particular instance where `Partially done` -> `WIP` I really meant something like: it partially addresses the concern but it is done. For example the diff testing _does_ test many parts of the staking module changes but maybe not all, and I dont' currently have plans to extend that.

Comment (danwt) [reply docs/quality_assurance.md:61]: Same for all places where I wrote `Partially done`.

Comment (danwt) [reply docs/quality_assurance.md:61]: Yeah sure, that's better!

File: docs/quality_assurance.md (hunk 1 above)
Comment (sainoe) [reply docs/quality_assurance.md:23]: LGTM

File: docs/quality_assurance.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+| Changes in IBC-GO testing suite | `Scheduled` (ibc-go team) | `NA` | `Partial coverage` | `NA` | `NA` |
+
+## Integration with Cosmos SDK
+
+A prerequisite of the code review is to open a PR with all the [SDK changes](https://github.com/cosmos/cosmos-sdk/tree/interchain-security-rebase) needed by Interchain Security.
+
+| Concern | Code Review | Unit Testing | Diff. testing | Testnet | Protocol audit |
+| ------- | ----------- | ------------ | ------------- | ------- | ----- |
+| Changes to staking module | `Scheduled` (sdk team) | `Partial coverage` <br /> see [unbonding_test.go](../x/ccv/provider/unbonding_test.go) <br />  redelegation and validator unbonding missing | `Partial coverage` | `Scheduled` | `NA` |
+| Changes to slashing module | `Scheduled` (sdk team) | `??` | `NA` | `Scheduled` | `NA` |
```
Comment (sainoe) [inline docs/quality_assurance.md:62]:
```suggestion
| Changes to slashing module | `Scheduled` (sdk team) | `covered` <br /> see [TestValidatorDowntime](../x/ccv/consumer/keeper/keeper_test.go#L345) <br />  | `NA` | `Scheduled` | `NA` |
```

File: docs/quality_assurance.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+| 2.09 | ICS-20 transfer | `Scheduled` (ibc-go team) | `??` | `NA` | `Scheduled` | 
+| 2.10 | Changes in IBC-GO testing suite | `Scheduled` (ibc-go team) | `NA` | `Partial coverage` | `NA` | 
+
+### Integration with Cosmos SDK
+
+A prerequisite of the code review is to open a PR with all the [SDK changes](https://github.com/cosmos/cosmos-sdk/tree/interchain-security-rebase) needed by Interchain Security.
+
+| ID | Concern | Code Review | Unit Testing | Diff. testing | Testnet | 
+| -- | ------- | ----------- | ------------ | ------------- | ------- | 
+| 3.01 | Changes to staking module | `Scheduled` (sdk team) | `Partial coverage` <br /> see [unbonding_test.go](../x/ccv/provider/unbonding_test.go) <br />  redelegation and validator unbonding missing | `Partial coverage` | `Scheduled` |
```
Comment (jtremback) [inline docs/quality_assurance.md:63]: There are also tests in the Cosmos SDK repo, should they be included here?

File: docs/quality_assurance.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+| ID | Concern | Code Review | Unit Testing | Diff. testing | Testnet | Protocol audit |
+| -- | ------- | ----------- | ------------ | ------------- | ------- | ----- |
+| 4.01 | Liveness of undelegations <br /> - unbonding delegation entries are eventually removed from `UnbondingDelegation` | `Scheduled` | `Done` <br /> see [unbonding_test.go](../x/ccv/provider/unbonding_test.go) | `Done` | `Scheduled` | `Scheduled` |
+| 4.02 | Liveness of redelegations <br /> - redelegations entries are eventually removed from `Redelegations` | `Scheduled` | `Scheduled` | `Scheduled` | `Scheduled` | `Scheduled` |
+| 4.03 | Liveness of validator unbondings <br /> - unbonding validators with no delegations are eventually removed from `Validators` | `Scheduled` | `Scheduled` | `Done` | `Scheduled` | `Scheduled` |
+| 4.04 | Unbonding operations (undelegations, redelegations, validator unbondings) should eventually complete even if the CCV channel is never established (due to error) <br /> - expected outcome: the pending VSC packets eventually timeout, which leads to the consumer chain removal | `Scheduled` | `??` | `Future work` | `Scheduled` | `Scheduled` <br /> high priority |
```
Comment (jtremback) [inline docs/quality_assurance.md:79]: This is an interesting one. What's the longest possible unbonding delay here? A way to avoid this altogether would be to have an ack from the consumer chain that a validator update had been received and not stop unbonding until then. That would be a lot of work though

**Review by jtremback — APPROVED** [review]

File: docs/quality_assurance.md (hunk 8 above)
Comment (mpoke) [reply docs/quality_assurance.md:63]: That would be a good idea.

File: docs/quality_assurance.md (hunk 9 above)
Comment (mpoke) [reply docs/quality_assurance.md:79]: > A way to avoid this altogether would be to have an ack from the consumer chain that a validator update had been received

How can a validator update be received without an established CCV channel?
